server/src/main/java/org/elasticsearch/common/bytes/BytesReferenceStreamInput.java

```diff
@@ -213,6 +215,23 @@ public void reset() throws IOException {
         }
     }
 
+    @Override
+    public BytesReference readBytesReference(int length) throws IOException {
+        if (length == 0) {
+            return BytesArray.EMPTY;
+        }
+        final List<BytesReference> slices = new ArrayList<>();
+        while (length > 0) {
+            maybeNextSlice();
+            final byte[] bytes = slice.array().clone();
```

Comment:
This looks inefficient, we might be copying a larger than necessary array here if we don't need the full slice?